### ISSUE TITLE
Non-default port for Cassandra is ignored.

### ISSUE THREAD

**Issue by pag-r** [hashicorp/vault#3467]
Hi,

I've noticed that when I try to use different port for Cassandra plugin (i.e. 9043), this port is ignored and the default one is used (9042). When I try to add the port number using notation hostname:port (node-0.cassandra3.host:9043), it always tries to connect to 9042 with error: Error creating session: no connections were made when creating the session.
The only solution was hardcoding port number in https://github.com/hashicorp/vault/blob/master/vendor/github.com/gocql/gocql/cluster.go#L137. 
Is there a chance to read port number from the hostname:port pair instead of changing the code itself and rebuilding vault?

**Comment by calvn** [comment]
Seems that you are getting into the same issue as https://github.com/hashicorp/vault/issues/3370. See [my comment](https://github.com/hashicorp/vault/issues/3370#issuecomment-331969756) on that thread regarding the issue with port parsing if provided as part of the host. You can pass in the port explicitly via 
https://www.vaultproject.io/api/secret/databases/cassandra.html#port.

**Comment by pag-r** [comment]
I've commented [gocql/gocql](https://github.com/gocql/gocql/issues/946) as this is directly connected to gocql not Vault itself. Adding the port even after the colon it ignores it.

**Comment by calvn** [comment]
Can you post the vault command that you are using? 

> Adding the port even after the colon ignores it.

The issue is that the port gets ignore when passing it this way, via as part of the host after the colon. By explicitly, I mean using the `port` parameter.

**Comment by calvn** [comment]
In your comment on https://github.com/gocql/gocql/issues/946#issuecomment-339659163, the port was passed in explicitly, just as if you were to do it by passing in the `port` parameter. The issue exists if you try the following:

```
package main

import (
	"fmt"
	"github.com/gocql/gocql"
)

func main() {
	cluster := gocql.NewCluster("127.0.0.1:9043")
	_, err := cluster.CreateSession()
	if err != nil {
		panic(err)
	}
	fmt.Println("cassandra init done")
}
```

**Comment by calvn** [comment]
I am sorry if I'm being confusing here - I think there was a misunderstand as to what providing the port "explicitly" meant. There is a bug/new behavior in the gocql library where the port is ignored if it's provided as part of the host (i.e. "127.0.0.1:9043"), which is not what I meant when I said "pass in the port explicitly". Instead, providing the port explicitly meant exactly how you provided the port in your comment example: https://github.com/gocql/gocql/issues/946#issuecomment-339659163. This same way of passing in the port, `cluster.Port = 9043`, is done in Vault via the `port` parameter.

In other words, perhaps try something along the lines of (replacing the parameters to ones that matches your current database configuration):
```
$ vault write database/config/cassandra \
    plugin_name=cassandra-database-plugin \
    allowed_roles="readonly" \
    hosts=localhost \
    port=9043 \
    username=cassandra \
    password=cassandra
```

**Comment by pag-r** [comment]
@calvn , I've got your point from the first time. And this is what I've set in config.json
This is how we I run Vault with params
```
 {
	"hosts":"node-0.cassandra.host", 
	"port":9043, 
	"username":"user",
	"password":"secret",
	"protocol_version":4
} 
```

Output:
```
Error creating session: gocql: unable to create session: control: unable to connect to initial hosts: dial tcp 10.255.6.168:9042:  getsockopt: connection refused
```

I've check the solution with adding port after the host, and still the same error
```
{
  "hosts":"node-0.cassandra.host:9043", 
  "username":"user",
  "password":"secret",
  "protocol_version":4
}
```
Error:
```
Error creating session: no connections were made when creating the session
```
When hardcoding port 9043 it's working properly

**Comment by calvn** [comment]
I see, what version of Vault are you running?

**Comment by pag-r** [comment]
Version 0.8.3

**Comment by calvn** [comment]
I am not able to reproduce the issue if using the port parameter:

With default 9042 port _(failure)_:
```
$ vault write database/config/cassandra plugin_name=cassandra-database-plugin allowed_roles="readonly" hosts=127.0.0.1 protocol_version=4 username=cassandra password=cassandra
Error writing data to database/config/cassandra: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/database/config/cassandra
Code: 400. Errors:

* error creating database object: error verifying connection: error creating session: gocql: unable to create session: control: unable to connect to initial hosts: dial tcp 127.0.0.1:9042: i/o timeout
```

With port as part of the host _(failure)_:
```
$ vault write database/config/cassandra plugin_name=cassandra-database-plugin allowed_roles="readonly" hosts=127.0.0.1:32779 protocol_version=4 username=cassandra password=cassandra
Error writing data to database/config/cassandra: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/database/config/cassandra
Code: 400. Errors:

* error creating database object: error verifying connection: error creating session: no connections were made when creating the session
```

With port parameter _(success)_:
```
$ vault write database/config/cassandra plugin_name=cassandra-database-plugin allowed_roles="readonly" hosts=127.0.0.1 port=32779 protocol_version=4 username=cassandra password=cassandra


The following warnings were returned from the Vault server:
* Read access to this endpoint should be controlled via ACLs as it will return the connection details as is, including passwords, if any.
```

Docker instance:
```
$ docker run --name some-cassandra -d -P -v `pwd`/cassandra/:/etc/cassandra/ -e CASSANDRA_BROADCAST_ADDRESS=127.0.0.1 cassandra
15f362bc44213bdca47b504ac733df1d01fb219ae543c8895e1c72de8ddef08b

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED                  STATUS              PORTS                                                                                                                         NAMES
15f362bc4421        cassandra           "/docker-entrypoin..."   Less than a second ago   Up 14 seconds       0.0.0.0:32782->7000/tcp, 0.0.0.0:32781->7001/tcp, 0.0.0.0:32780->7199/tcp, 0.0.0.0:32779->9042/tcp, 0.0.0.0:32778->9160/tcp   some-cassandra
```

The `` `pwd`/cassandra/cassandra.yaml`` file that gets mounted is simply the one from [test-fixtures/cassandra.yaml](https://github.com/hashicorp/vault/blob/master/plugins/database/cassandra/test-fixtures/cassandra.yaml) so that the correct authenticator is set.

**Comment by calvn** [comment]
Out of curiosity, which Cassandra database backend are you using? The [deprecated Cassandra backend](https://www.vaultproject.io/docs/secrets/cassandra/index.html) and the [new database Cassandra backed](https://www.vaultproject.io/docs/secrets/databases/cassandra.html) share the same gocql library, but the deprecated backend does not support port parameter, which might be why this is not working for you. 

If you're using the old backend, the recommendation here is to use the new database backend since you're on 0.8.3.

**Comment by pag-r** [comment]
We're using new backend, from some time.

**Comment by calvn** [comment]
Can you provide the vault CLI command/API request that you're using? The port parsing issue should be fixed upstream on https://github.com/gocql/gocql/pull/1009, but I'm curious at this point as to why your config request is not honoring the "port" param even with the new database backend.

**Comment by calvn** [comment]
@pag-r can Vault 0.9.0 has the gocql dep updated which should have resolved this. Can you report back here to determine if we are good on closing this issue?

**Comment by pag-r** [comment]
@calvn I will double check it once more, but when I've tested in just after release 0.9.0 it still didn't work without hardcoding. Will check it with pure new env and will get back to you with the results.

**Comment by jefferai** [comment]
Moving to 0.9.2 since we haven't heard back and the patch for gocql ought to be in 0.9.0.

**Comment by jefferai** [comment]
Closing due to lack of response.
